plugins/out_s3/s3.c

```diff
@@ -2050,42 +1938,132 @@ static void flush_init(void *out_context)
 
-        sched = flb_sched_ctx_get();
+    /*
+     * Set the target pipe channel: if this return code is running inside a
+     * thread pool worker, use the specific worker pipe/event loop to handle
+     * the return status, otherwise use the channel connected to the parent
+     * event loop.
+     */
+    if (flb_output_is_threaded(o_ins) == FLB_TRUE) {
+        /* Retrieve the thread instance and prepare pipe channel */
+        th_ins = flb_output_thread_instance_get();
+        pipe_fd = th_ins->ch_thread_events[1];
+    }
+    else {
+        pipe_fd = out_flush->o_ins->ch_events[1];
+    }
 
-        if (ctx->preserve_data_ordering) {
-            ret = flb_sched_timer_cb_create(sched, FLB_SCHED_TIMER_CB_PERM,
-                                            ctx->timer_ms, s3_upload_queue, ctx, NULL);
-        }
-        else {
-            ret = flb_sched_timer_cb_create(sched, FLB_SCHED_TIMER_CB_PERM,
-                                            ctx->timer_ms, cb_s3_upload, ctx, NULL);
-        }
-        if (ret == -1) {
-            flb_plg_error(ctx->ins, "Failed to create upload timer");
-            FLB_OUTPUT_RETURN(FLB_RETRY);
+    /* Notify the event loop about our return status */
+    n = flb_pipe_w(pipe_fd, (void *) &val, sizeof(val));
+    if (n == -1) {
+        flb_errno();
+    }
+}
+
+/* 
+ * We need to use async IO for S3 because it is more stable
+ * However, S3 has unique needs. The chunk list, multipart code, etc
+ * all is not concurrent safe. 
+ * Additionally, timer callbacks don't run in coroutines and
+ * can't using async IO. 
+ * Solution: daemon coroutine
+ * The first coroutine that that flushes to S3 never ends,
+ * and just uploads and sleeps. 
+ * 
+ * We increment the metrics counters for the chunk originally
+ * associated with the coroutine and decrement the task users
+ * in release_chunk_upstream()
+ */
+static void daemon_coroutine(struct flb_config *config, struct flb_s3 *ctx)
+{
+    flb_plg_info(ctx->ins, "May 2nd: daemon coroutine starting...");
+
+    ctx->daemon_coro_started = FLB_TRUE;
+
+    /* tell engine that this task did complete successfully */
+    flb_output_return_no_destroy(FLB_OK);
+
+    /* 
+     * FLB engine uses a graceful cooperative shutdown model. 
+     * If coroutines never end, the system won't stop.
+     * So the daemon coroutine must exit itself when the engine is in shutdown mode.
+     */
+    while (config->is_running == FLB_TRUE) {
+        /* Cleanup old buffers found on startup */
+        flush_startup_chunks(ctx);
+
+        /* upload any ready chunks */
+        cb_s3_upload(config, ctx);
+
+        if (config->is_running == FLB_FALSE) {
+            break;
         }
-        ctx->timer_created = FLB_TRUE;
+        /* 
+         * special coroutine sleep
+         * Doesn't block any thread
+         * Puts an event on the event
+         * loop which will wake this coro back up
+         */
+        flb_time_sleep(ctx->timer_ms);
```

Comment:
For the final version I think I may need to set this to be some static number that's always less than the `Grace` setting. Some fraction of it. Otherwise the sleep here can block shutdown until it resumes and eat up time that could be used to send data on shutdown.